warehouse/tracker.go

```diff
@@ -0,0 +1,183 @@
+package warehouse
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/timeutil"
+
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/services/stats"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/rudderlabs/rudder-server/warehouse/internal/model"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+func (wh *HandleT) CronTracker(ctx context.Context) error {
+	for {
+		wh.configSubscriberLock.RLock()
+
+		ch := make(chan warehouseutils.Warehouse, 1)
+		wg := sync.WaitGroup{}
+		wg.Add(len(wh.warehouses))
+
+		for _, warehouse := range wh.warehouses {
+			warehouse := warehouse
+
+			rruntime.GoForWarehouse(func() {
+				defer wg.Done()
+				ch <- warehouse
+			})
+		}
+		wh.configSubscriberLock.RUnlock()
+
+		rruntime.Go(func() {
+			wg.Wait()
+			close(ch)
+		})
+
+		for w := range ch {
+			if err := wh.Track(ctx, &w, config.Default); err != nil {
+				return fmt.Errorf("cron tracker failed for source: %s, destination: %s with error: %w", w.Source.ID, w.Destination.ID, err)
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			wh.Logger.Infof("context is cancelled, stopped running tracking")
+			return nil
+		case <-time.After(uploadStatusTrackFrequency):
+		}
+	}
+}
+
+func (wh *HandleT) Track(ctx context.Context, warehouse *warehouseutils.Warehouse, config *config.Config) error {
+	var (
+		query         string
+		queryArgs     []interface{}
+		createdAt     sql.NullTime
+		exists        bool
+		uploaded      int
+		syncFrequency = "1440"
+		Now           = timeutil.Now
+		NowSQL        = "NOW()"
+		timeWindow    = config.GetInt("Warehouse.uploadBufferTimeInMin", 180)
```

Comment:
Ideally, we should use time.Duration for `time.Window`, but this change might exceed the scope of this PR.

Reply:
Added this since we already have a fallback for when time unit is not present.